File: go/internal/thread/section.go
package thread

import (
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"regexp"
	"strings"
	"time"
)

// sectionRe matches section headers like "## Body", "## Notes", etc.
var sectionRe = regexp.MustCompile(`(?m)^## (\w+)`)

// hashCommentRe matches hash comments like "<!-- abc1 -->"
var hashCommentRe = regexp.MustCompile(`<!--\s*([a-f0-9]{4})\s*-->`)

// todoItemRe matches todo items like "- [ ] item" or "- [x] item"
var todoItemRe = regexp.MustCompile(`^- \[([ x])\] (.+?)\s*(<!--\s*[a-f0-9]{4}\s*-->)?$`)

// ExtractSection returns the content of a section (between ## Name and next ## or EOF)
func ExtractSection(content, name string) string {
	pattern := fmt.Sprintf(`(?ms)^## %s\n(.+?)(?:^## |\z)`, regexp.QuoteMeta(name))
	re := regexp.MustCompile(pattern)
	match := re.FindStringSubmatch(content)
	if len(match) < 2 {
		return ""
	}
	return strings.TrimSpace(match[1])
}

// ReplaceSection replaces the content of a section
func ReplaceSection(content, name, newContent string) string {
	pattern := fmt.Sprintf(`(?ms)(^## %s\n)(.+?)(^## |\z)`, regexp.QuoteMeta(name))
	re := regexp.MustCompile(pattern)

	if !re.MatchString(content) {
		// Section doesn't exist - handled by caller
		return content
	}

	return re.ReplaceAllString(content, fmt.Sprintf("${1}\n%s\n\n${3}", newContent))
}

// AppendToSection appends content to a section
func AppendToSection(content, name, addition string) string {
	sectionContent := ExtractSection(content, name)
	newContent := strings.TrimSpace(sectionContent)
	if newContent != "" {
		newContent += "\n"
	}
	newContent += addition
	return ReplaceSection(content, name, newContent)
}

// EnsureSection creates a section if it doesn't exist, placing it before another section
func EnsureSection(content, name, before string) string {
	pattern := fmt.Sprintf(`(?m)^## %s`, regexp.QuoteMeta(name))
	if regexp.MustCompile(pattern).MatchString(content) {
		return content
	}

	beforePattern := fmt.Sprintf(`(?m)(^## %s)`, regexp.QuoteMeta(before))
	beforeRe := regexp.MustCompile(beforePattern)

	if beforeRe.MatchString(content) {
		return beforeRe.ReplaceAllString(content, fmt.Sprintf("## %s\n\n$1", name))
	}

	// If before section doesn't exist, append at end
	return content + fmt.Sprintf("\n## %s\n\n", name)
}

// GenerateHash creates a 4-character hash for an item
func GenerateHash(text string) string {
	data := fmt.Sprintf("%s%d", text, time.Now().UnixNano())
	hash := md5.Sum([]byte(data))
	return hex.EncodeToString(hash[:])[:4]
}

// InsertLogEntry adds a timestamped entry to the Log section
func InsertLogEntry(content, entry string) string {
	today := time.Now().Format("2006-01-02")
	timestamp := time.Now().Format("15:04")
	bulletEntry := fmt.Sprintf("- **%s** %s", timestamp, entry)
	heading := fmt.Sprintf("### %s", today)

	// Check if today's heading exists
	todayPattern := regexp.MustCompile(fmt.Sprintf(`(?m)^### %s`, regexp.QuoteMeta(today)))
	if todayPattern.MatchString(content) {
		// Insert after today's heading
		pattern := fmt.Sprintf(`(?m)(^### %s\n)`, regexp.QuoteMeta(today))
		re := regexp.MustCompile(pattern)
		return re.ReplaceAllString(content, fmt.Sprintf("${1}\n%s\n", bulletEntry))
	}

	// Check if Log section exists
	logPattern := regexp.MustCompile(`(?m)^## Log`)
	if logPattern.MatchString(content) {
		// Insert new heading after ## Log
		return logPattern.ReplaceAllString(content, fmt.Sprintf("## Log\n\n%s\n\n%s", heading, bulletEntry))
	}

	// No Log section - append one
	return content + fmt.Sprintf("\n## Log\n\n%s\n\n%s\n", heading, bulletEntry)
}

// AddNote adds a note to the Notes section with a hash comment
func AddNote(content, text string) (string, string) {
	// Ensure Notes section exists
	content = EnsureSection(content, "Notes", "Todo")

	hash := GenerateHash(text)
	noteEntry := fmt.Sprintf("- %s  <!-- %s -->", text, hash)

	// Insert at top of Notes section
	pattern := regexp.MustCompile(`(?m)(^## Notes\n)`)
	newContent := pattern.ReplaceAllString(content, fmt.Sprintf("${1}\n%s\n", noteEntry))

	return newContent, hash
}

// RemoveByHash removes a line containing the specified hash comment from a section
func RemoveByHash(content, section, hash string) (string, error) {
	lines := strings.Split(content, "\n")
	inSection := false
	hashPattern := fmt.Sprintf("<!-- %s", hash)
	found := false

	var result []string
	for _, line := range lines {
		if strings.HasPrefix(line, "## "+section) {
			inSection = true
		} else if strings.HasPrefix(line, "## ") {
			inSection = false
		}

		if inSection && strings.Contains(line, hashPattern) && !found {
			found = true
			continue // skip this line
		}
		result = append(result, line)
	}

	if !found {
		return content, fmt.Errorf("no item with hash '%s' found", hash)
	}

	return strings.Join(result, "\n"), nil
}

// EditByHash replaces the text of an item by hash
func EditByHash(content, section, hash, newText string) (string, error) {
	lines := strings.Split(content, "\n")
	inSection := false
	hashPattern := fmt.Sprintf("<!-- %s", hash)
	found := false

	var result []string
	for _, line := range lines {
		if strings.HasPrefix(line, "## "+section) {
			inSection = true
		} else if strings.HasPrefix(line, "## ") {
			inSection = false
		}

		if inSection && strings.Contains(line, hashPattern) && !found {
			found = true
			// Extract hash from line and rebuild
			match := hashCommentRe.FindStringSubmatch(line)
			if len(match) > 1 {
				result = append(result, fmt.Sprintf("- %s  <!-- %s -->", newText, match[1]))
				continue
			}
		}
		result = append(result, line)
	}

	if !found {
		return content, fmt.Errorf("no item with hash '%s' found", hash)
	}

	return strings.Join(result, "\n"), nil
}

// AddTodoItem adds a checkbox item to the Todo section
func AddTodoItem(content, text string) (string, string) {
	hash := GenerateHash(text)
	todoEntry := fmt.Sprintf("- [ ] %s  <!-- %s -->", text, hash)

	// Insert at top of Todo section
	pattern := regexp.MustCompile(`(?m)(^## Todo\n)`)
	newContent := pattern.ReplaceAllString(content, fmt.Sprintf("${1}\n%s\n", todoEntry))

	return newContent, hash
}

// SetTodoChecked sets a todo item's checked state by hash
func SetTodoChecked(content, hash string, checked bool) (string, error) {
	lines := strings.Split(content, "\n")
	inTodo := false
	hashPattern := fmt.Sprintf("<!-- %s", hash)
	found := false

	var result []string
	for _, line := range lines {
		if strings.HasPrefix(line, "## Todo") {
			inTodo = true
		} else if strings.HasPrefix(line, "## ") {
			inTodo = false
		}

		if inTodo && strings.Contains(line, hashPattern) && !found {
			found = true
			if checked {
				line = strings.Replace(line, "- [ ]", "- [x]", 1)
			} else {
				line = strings.Replace(line, "- [x]", "- [ ]", 1)
			}
		}
		result = append(result, line)
	}

	if !found {
		return content, fmt.Errorf("no item with hash '%s' found", hash)
	}

	return strings.Join(result, "\n"), nil
}

// CountMatchingItems counts items matching a hash prefix in a section
func CountMatchingItems(content, section, hash string) int {
	lines := strings.Split(content, "\n")
	inSection := false
	hashPattern := fmt.Sprintf("<!-- %s", hash)
	count := 0

	for _, line := range lines {
		if strings.HasPrefix(line, "## "+section) {
			inSection = true
		} else if strings.HasPrefix(line, "## ") {
			inSection = false
		}

		if inSection && strings.Contains(line, hashPattern) {
			count++
		}
	}

	return count
}

File: swift/Sources/threads/Section.swift
import Foundation

// extractSection returns the content of a section (between ## Name and next ## or EOF)
func extractSection(_ content: String, _ name: String) -> String {
    let pattern = "(?ms)^## \(NSRegularExpression.escapedPattern(for: name))\n(.+?)(?:^## |\\z)"
    guard let regex = try? NSRegularExpression(pattern: pattern, options: [.anchorsMatchLines, .dotMatchesLineSeparators]),
          let match = regex.firstMatch(in: content, range: NSRange(content.startIndex..., in: content)),
          let range = Range(match.range(at: 1), in: content) else {
        return ""
    }
    return String(content[range]).trimmingCharacters(in: .whitespacesAndNewlines)
}

// replaceSection replaces the content of a section
func replaceSection(_ content: String, _ name: String, _ newContent: String) -> String {
    let pattern = "(?ms)(^## \(NSRegularExpression.escapedPattern(for: name))\n)(.+?)(^## |\\z)"
    guard let regex = try? NSRegularExpression(pattern: pattern, options: [.anchorsMatchLines, .dotMatchesLineSeparators]),
          let match = regex.firstMatch(in: content, range: NSRange(content.startIndex..., in: content)) else {
        // Section doesn't exist
        return content
    }

    // Use functional replacement to avoid $ escaping issues in templates
    guard let headerRange = Range(match.range(at: 1), in: content),
          let tailRange = Range(match.range(at: 3), in: content) else {
        return content
    }

    let header = String(content[headerRange])
    let tail = String(content[tailRange])
    let fullMatchRange = Range(match.range, in: content)!

    var result = content
    result.replaceSubrange(fullMatchRange, with: "\(header)\n\(newContent)\n\n\(tail)")
    return result
}

// appendToSection appends content to a section
func appendToSection(_ content: String, _ name: String, _ addition: String) -> String {
    let sectionContent = extractSection(content, name)
    var newContent = sectionContent.trimmingCharacters(in: .whitespacesAndNewlines)
    if !newContent.isEmpty {
        newContent += "\n"
    }
    newContent += addition
    return replaceSection(content, name, newContent)
}

// ensureSection creates a section if it doesn't exist, placing it before another section
func ensureSection(_ content: String, _ name: String, _ before: String) -> String {
    let pattern = "(?m)^## \(NSRegularExpression.escapedPattern(for: name))"
    if let regex = try? NSRegularExpression(pattern: pattern),
       regex.firstMatch(in: content, range: NSRange(content.startIndex..., in: content)) != nil {
        return content // Section already exists
    }

    let beforePattern = "(?m)(^## \(NSRegularExpression.escapedPattern(for: before)))"
    if let beforeRegex = try? NSRegularExpression(pattern: beforePattern),
       beforeRegex.firstMatch(in: content, range: NSRange(content.startIndex..., in: content)) != nil {
        return beforeRegex.stringByReplacingMatches(
            in: content,
            range: NSRange(content.startIndex..., in: content),
            withTemplate: "## \(name)\n\n$1"
        )
    }

    // If before section doesn't exist, append at end
    return content + "\n## \(name)\n\n"
}

// generateHash creates a 4-character hash for an item using FNV-1a
func generateHash(_ text: String) -> String {
    let data = "\(text)\(Date().timeIntervalSince1970)"
    var hash: UInt64 = 14695981039346656037  // FNV offset basis
    for byte in data.utf8 {
        hash ^= UInt64(byte)
        hash &*= 1099511628211  // FNV prime
    }
    // Extract 2 bytes (4 hex chars)
    return String(format: "%04x", hash & 0xFFFF)
}

// insertLogEntry adds a timestamped entry to the Log section
func insertLogEntry(_ content: String, _ entry: String) -> String {
    let dateFormatter = DateFormatter()
    dateFormatter.dateFormat = "yyyy-MM-dd"
    let today = dateFormatter.string(from: Date())

    dateFormatter.dateFormat = "HH:mm"
    let timestamp = dateFormatter.string(from: Date())

    let bulletEntry = "- **\(timestamp)** \(entry)"
    let heading = "### \(today)"

    // Check if today's heading exists - use functional replacement to preserve $ in entry
    let todayPattern = "(?m)(^### \(NSRegularExpression.escapedPattern(for: today))\n)"
    if let todayRegex = try? NSRegularExpression(pattern: todayPattern),
       let match = todayRegex.firstMatch(in: content, range: NSRange(content.startIndex..., in: content)),
       let matchRange = Range(match.range, in: content) {
        var result = content
        let matched = String(content[matchRange])
        result.replaceSubrange(matchRange, with: "\(matched)\n\(bulletEntry)\n")
        return result
    }

    // Check if Log section exists
    if let match = CachedRegex.logSection.firstMatch(in: content, range: NSRange(content.startIndex..., in: content)),
       let matchRange = Range(match.range, in: content) {
        var result = content
        result.replaceSubrange(matchRange, with: "## Log\n\n\(heading)\n\n\(bulletEntry)")
        return result
    }

    // No Log section - append one
    return content + "\n## Log\n\n\(heading)\n\n\(bulletEntry)\n"
}

// addNote adds a note to the Notes section with a hash comment
func addNote(_ content: String, _ text: String) -> (String, String) {
    // Ensure Notes section exists
    var newContent = ensureSection(content, "Notes", "Todo")

    let hash = generateHash(text)
    let noteEntry = "- \(text)  <!-- \(hash) -->"

    // Insert at top of Notes section - use functional replacement to preserve $ in text
    let pattern = "(?m)(^## Notes\n)"
    if let regex = try? NSRegularExpression(pattern: pattern),
       let match = regex.firstMatch(in: newContent, range: NSRange(newContent.startIndex..., in: newContent)),
       let matchRange = Range(match.range, in: newContent) {
        let matched = String(newContent[matchRange])
        newContent.replaceSubrange(matchRange, with: "\(matched)\n\(noteEntry)\n")
    }

    return (newContent, hash)
}

// removeByHash removes a line containing the specified hash comment from a section
func removeByHash(_ content: String, _ section: String, _ hash: String) throws -> String {
    let lines = content.components(separatedBy: "\n")
    var inSection = false
    let hashPattern = "<!-- \(hash)"
    var found = false

    var result: [String] = []
    for line in lines {
        if line.hasPrefix("## \(section)") {
            inSection = true
        } else if line.hasPrefix("## ") {
            inSection = false
        }

        if inSection && line.contains(hashPattern) && !found {
            found = true
            continue // skip this line
        }
        result.append(line)
    }

    if !found {
        throw SectionError.hashNotFound(hash)
    }

    return result.joined(separator: "\n")
}

// editByHash replaces the text of an item by hash
func editByHash(_ content: String, _ section: String, _ hash: String, _ newText: String) throws -> String {
    let lines = content.components(separatedBy: "\n")
    var inSection = false
    let hashPattern = "<!-- \(hash)"
    var found = false

    var result: [String] = []

    for line in lines {
        if line.hasPrefix("## \(section)") {
            inSection = true
        } else if line.hasPrefix("## ") {
            inSection = false
        }

        if inSection && line.contains(hashPattern) && !found {
            found = true
            // Extract hash from line and rebuild
            if let match = CachedRegex.hashComment.firstMatch(in: line, range: NSRange(line.startIndex..., in: line)),
               let range = Range(match.range(at: 1), in: line) {
                let extractedHash = String(line[range])
                result.append("- \(newText)  <!-- \(extractedHash) -->")
                continue
            }
        }
        result.append(line)
    }

    if !found {
        throw SectionError.hashNotFound(hash)
    }

    return result.joined(separator: "\n")
}

// addTodoItem adds a checkbox item to the Todo section
func addTodoItem(_ content: String, _ text: String) -> (String, String) {
    let hash = generateHash(text)
    let todoEntry = "- [ ] \(text)  <!-- \(hash) -->"

    // Insert at top of Todo section - use functional replacement to preserve $ in text
    let pattern = "(?m)(^## Todo\n)"
    if let regex = try? NSRegularExpression(pattern: pattern),
       let match = regex.firstMatch(in: content, range: NSRange(content.startIndex..., in: content)),
       let matchRange = Range(match.range, in: content) {
        var newContent = content
        let matched = String(content[matchRange])
        newContent.replaceSubrange(matchRange, with: "\(matched)\n\(todoEntry)\n")
        return (newContent, hash)
    }

    return (content, hash)
}

// setTodoChecked sets a todo item's checked state by hash
func setTodoChecked(_ content: String, _ hash: String, _ checked: Bool) throws -> String {
    let lines = content.components(separatedBy: "\n")
    var inTodo = false
    let hashPattern = "<!-- \(hash)"
    var found = false

    var result: [String] = []
    for var line in lines {
        if line.hasPrefix("## Todo") {
            inTodo = true
        } else if line.hasPrefix("## ") {
            inTodo = false
        }

        if inTodo && line.contains(hashPattern) && !found {
            found = true
            if checked {
                line = line.replacingOccurrences(of: "- [ ]", with: "- [x]")
            } else {
                line = line.replacingOccurrences(of: "- [x]", with: "- [ ]")
            }
        }
        result.append(line)
    }

    if !found {
        throw SectionError.hashNotFound(hash)
    }

    return result.joined(separator: "\n")
}

// countMatchingItems counts items matching a hash prefix in a section
func countMatchingItems(_ content: String, _ section: String, _ hash: String) -> Int {
    let lines = content.components(separatedBy: "\n")
    var inSection = false
    let hashPattern = "<!-- \(hash)"
    var count = 0

    for line in lines {
        if line.hasPrefix("## \(section)") {
            inSection = true
        } else if line.hasPrefix("## ") {
            inSection = false
        }

        if inSection && line.contains(hashPattern) {
            count += 1
        }
    }

    return count
}

enum SectionError: Error, LocalizedError {
    case hashNotFound(String)

    var errorDescription: String? {
        switch self {
        case .hashNotFound(let hash):
            return "no item with hash '\(hash)' found"
        }
    }
}
